Added messaging for multiple users and new feature user's group

captjt opened the pull request:
* `go fmt` on DiscardStats methods

captjt committed:
go fmt on DiscardStats methods

captjt committed:
added multiuser messaging
also fixed some formatting and starting working on getting the group messsaging in place

captjt committed:
more idiomatic imports and misc.

captjt committed:
setup inital groups structure

briankassouf commented on .gitignore:
```diff
@@ -1 +1,2 @@
 *.sw[o-p]
+vendor
```
We don't want to ignore the vendor directory. Eventually (in another PR) all the vendored packages should be updated. I think this will require some code updates too, but shouldn't be too bad.

briankassouf commented on Dockerfile:
```diff
@@ -1,9 +1,8 @@
-ADD . /go/src/github.com/Imgur/incus
-WORKDIR /go/src/github.com/Imgur/incus
+ADD . /go/src/github.com/jtaylor32/incus
+WORKDIR /go/src/github.com/jtaylor32/incus
```
Can you be sure to change this back? When working on a fork with Golang I find this handy: https://stackoverflow.com/questions/14323872/using-forked-package-import-in-go

captjt replied:
Perfect great source. Thanks!

briankassouf commented on README.md:
```diff
@@ -213,7 +213,7 @@ sudo /etc/init.d/incus restart
 Incus needs to be restarted after any configuration change.
 
-#### CLIENT_BROADCASTS
+#### ClientBroadcast
```
This should be changed back as the configuration option should still be the underscore'd version. see config.go for more info.

briankassouf commented on glide.lock:
```diff
@@ -0,0 +1,86 @@
+  - unicode/norm
+- name: gopkg.in/yaml.v2
+  version: a3f3340b5840cee44f372bddb5880fcbc419b46a
+testImports: []
```
The current vendor tool is Godeps. Any major vendor changes should be done in another PR. (Anything that doesn't pertain to the groups feature)

captjt replied:
@briankassouf do you have any preference whether or not I open a PR on using a different tool? This raising the questions of how you'd like to handle the vendor directory as well. We can address this after I clean up this PR.

briankassouf commented on incus/main.go:
```diff
@@ -11,7 +11,7 @@ import (
 	"time"
 
-	"github.com/Imgur/incus"
+	"github.com/jtaylor32/incus"
```
Same, please make sure this is changed back

briankassouf commented on scripts/install.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-GOPROJ=$GOPATH/src/github.com/Imgur/incus
+GOPROJ=$GOPATH/src/github.com/jtaylor32/incus
```
Same as above.

briankassouf commented on store.go:
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
 }
+
+func (this *Storage) SetGroups(sock *Socket) error {
+	this.pageMu.Lock()
```
Can you add a new lock to the storage struct? It should use a separate lock for the groups map.

briankassouf commented on store.go:
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
+}
+
+func (this *Storage) UnsetGroups(sock *Socket) error {
+	this.pageMu.Lock()
```
Same, it should use a different lock.

briankassouf commented on store.go:
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
+	this.pageMu.Unlock()
+
+	if this.StorageType == "redis" {
+		if err := this.redis.UnsetPage(sock); err != nil {
```
`UnsetGroup`

briankassouf commented on store.go:
```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
+}
+
+func (this *Storage) getGroups(groups []string) []map[string]*Socket {
+	defer this.pageMu.RUnlock()
```
Same, should use a different lock. Also could you move the defer statement below the `Lock()` call?

briankassouf commented on message.go:
```diff
@@ -338,6 +375,74 @@ func (this *CommandMsg) messageUser(UID string, page string, server *Server) {
+			}
+		}
+		return nil
+	})
```
Rather than spawn two more go routines, what if it looped through the users and called `this.messageUser(uid, page, server)`?

captjt replied:
My thought process here was messageUsers will actually take a single User as well similar to messageUser (removing the messageUser function). The messaging process with the concurrency of two goroutines seems nicer at scale if someone was sending many-many messages with Incus. If you think that a simple loop would be faster I can change that as well.

briankassouf commented on message.go:
```diff
@@ -380,7 +485,31 @@ func (this *CommandMsg) messagePage(page string, server *Server) {
+
+	server.Stats.LogPageMessage()
+
+	groupMap := server.Store.getGroups(groups)
```
This is technically a `groupList`

briankassouf commented on message.go:
```diff
@@ -61,6 +63,22 @@ func (this *CommandMsg) FromSocket(sock *Socket) {
+		}
+
+		// hack check for empty slice
+		if len(sock.Groups) != 0 {
```
Should this be moved above the `len(groups) == 0` so that an empty group command could remove all the socket from all groups?

captjt replied:
What I was trying to do is add groups to a single socket. So let's say you have your UID and you are apart of 3 different groups. I have mine and I'm in 2 groups. Someone sends you a message and to one of my groups. We both get that message kind of thing. Without too much knowledge of Redis I've been messing with just clearing them when setting new groups. Will gladly take recommendations on how you'd implement it.

briankassouf commented on stats.go:
```diff
@@ -39,7 +40,7 @@ type DiscardStats struct{}
+func (d *DiscardStats) LogGoroutines(int)                             {}
 func (d *DiscardStats) LogCommand(from, cmdType string)               {}
 func (d *DiscardStats) LogPageMessage()                               {}
 func (d *DiscardStats) LogUserMessage()                               {}
```
Could you add a `LogGroupMessage()` stat function?

captjt replied:
Absolutely. I apologize at first it was only the initial formatting. Just kept adding stuff while I went along. I will clean everything and make sure it's more clear of what changes where made. 

Thanks!

briankassouf commented on message.go:
```diff
@@ -380,7 +485,31 @@ func (this *CommandMsg) messagePage(page string, server *Server) {
+		return
+	}
+
+	server.Stats.LogPageMessage()
```
This should be calling a `LogGroupMessage()` function and should probably take in the length of the group array. so something like: `LogGroupMessage(len(groups))`

captjt replied:
I will get the logging function finished and pushed up today.

briankassouf reviewed (changes requested):
Overall really awesome! Just a few minor things and nitpicks. Thanks for adding this feature!

briankassouf commented:
Also could you update the title of the PR to reflect the added features? And look into adding some test cases for the groups feature?

captjt committed:
added groups mutex and moved defers to after locks

captjt committed:
added group logging for group count
also moved some defer calls to after the mutex locks
and some more formatting cleanup/revert back to Imgur stuff

jtaylor32 commented:
Will rebase and squash to clean up git history. 🙈

captjt committed:
removed glide tool files

captjt committed:
removed the "groups" concept
converted multiple user's groups to a single group similar to a "page"

captjt commented:
I removed the "groups" concept for right now. User's can have a single _group_ similar to a _page_. This is because it will be a bigger feature for a different PR. If you have a group of friends you want to chat with while that's separate from page-to-page you can set a "group" for chatting that way.

captjt commented:
@briankassouf with what I made some changes to what else do you think needs to be tweaked?
